versioned/persist/tx/src/main/java/org/projectnessie/versioned/persist/tx/TxDatabaseAdapter.java

```diff
@@ -1684,4 +1709,129 @@ private void writeRefLog(Connection connection, AdapterTypes.RefLogEntry entry)
+        int paramIndex = 2;
+        Map<String, ContentAttachmentKey> keyMap = new HashMap<>();
+        for (ContentAttachmentKey key : keysList) {
+          String keyAsString = attachmentKeyAsString(key);
+          keyMap.put(keyAsString, key);
+          ps.setString(paramIndex++, keyAsString);
+        }
+        List<ContentAttachment> result = new ArrayList<>();
+        try (ResultSet rs = ps.executeQuery()) {
+          while (rs.next()) {
+            String id = rs.getString(1);
+            byte[] value = rs.getBytes(2);
+            ContentAttachmentKey key = keyMap.get(id);
+            result.add(attachmentContent(key, parseFrom(value)));
+          }
+        }
+        return result.stream();
+      }
+    } catch (SQLException | InvalidProtocolBufferException e) {
+      throw new RuntimeException(e);
+    } finally {
+      releaseConnection(conn);
+    }
+  }
+
+  @Override
+  public void putAttachments(Stream<ContentAttachment> attachments) {
+    Connection conn = borrowConnection();
+    try {
+      try (PreparedStatement ps = conn.prepareStatement(SqlStatements.INSERT_ATTACHMENT)) {
+        attachments.forEach(
+            bc -> {
+              try {
+                attachmentInsert(ps, bc);
+              } catch (SQLException e) {
+                throw new RuntimeException(e);
+              }
+            });
+      }
+      txCommit(conn);
+    } catch (SQLException e) {
+      throw new RuntimeException(e);
+    } finally {
+      releaseConnection(conn);
+    }
+  }
+
+  private void attachmentInsert(PreparedStatement ps, ContentAttachment attachment)
+      throws SQLException {
+    ps.setString(1, config.getRepositoryId());
+    ps.setString(2, attachmentKeyAsString(ContentAttachmentKey.fromContent(attachment)));
+    ps.setBytes(3, ProtoSerialization.toProtoValue(attachment).toByteArray());
+    ps.setString(4, attachment.getVersion());
+    ps.executeUpdate();
+  }
+
+  @Override
+  public boolean consistentPutAttachment(
+      ContentAttachment attachment, Optional<String> expectedVersion) {
+    Connection conn = borrowConnection();
+    try {
+      if (expectedVersion.isPresent()) {
+        try (PreparedStatement ps = conn.prepareStatement(SqlStatements.UPDATE_ATTACHMENT)) {
+          ps.setBytes(1, ProtoSerialization.toProtoValue(attachment).toByteArray());
+          ps.setString(2, attachment.getVersion());
+          ps.setString(3, config.getRepositoryId());
+          ps.setString(4, attachmentKeyAsString(ContentAttachmentKey.fromContent(attachment)));
+          ps.setString(5, expectedVersion.get());
+          if (ps.executeUpdate() != 1) {
+            return false;
+          }
+        }
+      } else {
+        try (PreparedStatement ps = conn.prepareStatement(SqlStatements.INSERT_ATTACHMENT)) {
+          attachmentInsert(ps, attachment);
+        }
+      }
+      txCommit(conn);
+    } catch (SQLException e) {
+      if (isIntegrityConstraintViolation(e)) {
+        return false;
```

Comment:
shouldn't we log that? Right now the information about the failure will be hidden (because we are just returning `false`)